src/Stratis.Bitcoin/Consensus/ChainedHeaderTree.cs

```diff
@@ -113,11 +134,331 @@ public void Initialize(ChainedHeader consensusTip)
+                }
+            }
+            
+            this.logger.LogTrace("(-):*.{0}={1},{2}={3}", nameof(headersToValidate.Count), headersToValidate.Count, nameof(reorgRequired), reorgRequired);
+            return headersToValidate;
+        }
+
+        /// <summary>Sets the tip claim for a peer.</summary>
+        /// <param name="networkPeerId">Peer Id.</param>
+        /// <param name="tipHash">Tip's hash.</param>
+        private void ClaimPeerTip(int networkPeerId, uint256 tipHash)
+        {
+            this.logger.LogTrace("({0}:{1},{2}:'{3}')", nameof(networkPeerId), networkPeerId, nameof(tipHash), tipHash);
+            
+            HashSet<int> peersClaimingThisHeader;
+            if (!this.peerIdsByTipHash.TryGetValue(tipHash, out peersClaimingThisHeader))
+            {
+                peersClaimingThisHeader = new HashSet<int>();
+                this.peerIdsByTipHash.Add(tipHash, peersClaimingThisHeader);
+            }
+
+            peersClaimingThisHeader.Add(networkPeerId);
+           
+            this.logger.LogTrace("(-)");
+        }
+
+        /// <summary>
+        /// Handles situation when block data was considered to be invalid
+        /// for a given header during the partial or full validation.
+        /// </summary>
+        /// <param name="chainedHeader">Chained header which block data failed the validation.</param>
+        /// <returns>List of peer Ids that were claiming chain that contains an invalid block. Such peers should be banned.</returns>
+        public List<int> PartialOrFullValidationFailed(ChainedHeader chainedHeader)
+        {
+            this.logger.LogTrace("({0}:'{1}')", nameof(chainedHeader), chainedHeader);
+
+            List<int> peersToBan = this.RemoveSubtree(chainedHeader);
+
+            this.RemoveUnclaimedBranch(chainedHeader.Previous);
+
+            this.logger.LogTrace("(-):*.{0}={1}", nameof(peersToBan.Count), peersToBan.Count);
+            return peersToBan;
+        }
+
+        /// <summary>
+        /// Handles situation when consensuses tip was changed.
+        /// </summary>
+        /// <param name="newConsensusTip">The new consensus tip.</param>
+        /// <returns>List of peer Ids that violate max reorg rule.</returns>
+        public List<int> ConsensusTipChanged(ChainedHeader newConsensusTip)
+        {
+            this.logger.LogTrace("({0}:'{1}')", nameof(newConsensusTip), newConsensusTip);
+
+            // Switch consensus tip to the new block header.
+            this.AddOrReplacePeerTip(LocalPeerId, newConsensusTip.HashBlock);
+
+            List<int> peerIdsToResync = this.FindPeersToResync(newConsensusTip);
+
+            // Remove block data for the headers that are too far from the consensus tip.
+            this.CleanOldBlockDataFromMemory(newConsensusTip);
+
+            this.logger.LogTrace("(-)");
+            return peerIdsToResync;
+        }
+
+        /// <summary>Checks each peer's tip if it violates max reorg rule.</summary>
```

Comment:
The description does not sound correct (or the method name is not correct)

Reply:
Fixed